### PR TITLE
Server doesn't receive a channelInactive() from its client

### PR DESCRIPTION
**Pull request by rkapsi**
Server doesn't receive a channelInactive() from its client after it's turning off auto reading.

I started to write a new set of Unit Tests for #5081 and came across this problem. The testDisconnectToggeAutoRead test should fail.

### CODE REVIEW

**Comment by normanmaurer** [comment]
@rkapsi thanks for reporting and for the test-case.

**Comment by normanmaurer** [comment]
@rkapsi also are we allowed to re-use some code from the test-case ?

**Comment by rkapsi** [comment]
@normanmaurer yes

**Comment by rkapsi** [comment]
@normanmaurer I just noticed that I was setting `.option(ChannelOption.AUTO_READ, autoRead)` on the server side which needs to be `.childOption()`. Now both tests are failing.

**Comment by normanmaurer** [comment]
@rkapsi is this something that worked before ?

**Comment by rkapsi** [comment]
@normanmaurer I don't know but I hope it did. I stumbled onto it when I was writing a completely differently unit test where I was trying to trigger something off the `channelInactive(...)` event.

File: transport/src/test/java/io/netty/channel/ChannelCloseTest.java [hunk 1]
```diff
@@ -0,0 +1,257 @@
...
+    /**
+     * Scenario 1
+     *
+     * 1. The server starts with auto reading turned on
+     * 2. The client connects
+     * 3. The client sends a message
+     * 4. The server receives the message
+     * 5. The server turns off auto reading
+     * 6. The client disconnects
+     * 7. The server (should) receive an channelInactive() event
```
Comment (Scottmitch) [inline transport/src/test/java/io/netty/channel/ChannelCloseTest.java]: I think the only way to detect the remote peer has closed (assuming the close follows the TCP protocol) in java NIO is to set the OP_READ or OP_WRITE option on the SelectionKey (assuming you have already connected). By turning off autoRead this should clear the OP_READ on SelectionKey (see https://github.com/netty/netty/pull/5101), and we only set OP_WRITE if we are not able to write to the socket any more. I can provide a "raw NIO" example which demonstrates this behavior if necessary.

Comment (normanmaurer) [inline transport/src/test/java/io/netty/channel/ChannelCloseTest.java]: +1

Comment (trustin) [inline transport/src/test/java/io/netty/channel/ChannelCloseTest.java]: Yes, there's no way to get notified when a peer closed the stream without OP_READ set.

**Comment by normanmaurer** [comment]
So let me close this as "not a bug"

**Comment by rkapsi** [comment]
Duh, this makes a lot sense. Sorry about the scare.

File: transport/src/test/java/io/netty/channel/ChannelCloseTest.java (hunk 1 above)
Comment (johnou) [inline transport/src/test/java/io/netty/channel/ChannelCloseTest.java]: @Scottmitch would it theoretically be possible for me to miss out on channelInactive events if I set auto read in this manner?

```java
    private void sendToClient(ByteBuf buffer, boolean closeAfterWrite) {
        if (closeAfterWrite) {
            channel.config().setAutoRead(false);
            ChannelFuture future = channel.writeAndFlush(buffer);
            future.addListener(ChannelFutureListener.CLOSE);
        } else {
            channel.writeAndFlush(buffer, channel.voidPromise());
        }
    }
```

Comment (normanmaurer) [inline transport/src/test/java/io/netty/channel/ChannelCloseTest.java]: @johnou nope as you add a future that explicit close the channel.

**Pull request closed without merging**